Vulkan build on MacOS

dokterbob opened the pull request:
Significant (~30%) performance boost on Intel Mac. 

I'm getting up to ~10 toks/s with 28 layers offloaded to a 4GB AMD Radeon Pro 5500M. Up from ~7 tok/s on just CPU.

dokterbob committed:
Allow for Vulkan build with Accelerate.

Closes #5304

dokterbob committed:
Resolve ErrorIncompatibleDriver with Vulkan on MacOS.

Refs:
- https://chat.openai.com/share/7020ce72-65fc-45ec-b7be-9d9d798a5f3f
- https://github.com/SaschaWillems/Vulkan/issues/954
- https://github.com/haasn/libplacebo/issues/128
- https://github.com/KhronosGroup/Vulkan-Samples/issues/476

0cc4m commented on ggml-vulkan.cpp:
```diff
@@ -978,8 +978,10 @@ void ggml_vk_init() {
     };
-    vk::InstanceCreateInfo instance_create_info(vk::InstanceCreateFlags(), &app_info, layers, extensions);
+
+    vk::InstanceCreateInfo instance_create_info(vk::InstanceCreateFlags(vk::InstanceCreateFlagBits::eEnumeratePortabilityKHR), &app_info, layers, extensions);
```
I'm gonna do a proper test tomorrow. But you should wrap these changes into preprocessor checks so that they only happen on Apple devices.

dokterbob replied:
I considered that, but then again if these features do not cause problems on other platforms, and it yields a more generic and simple solution, shouldn't we just leave them on?

Your call. Curious to see whether you're running into any issues on your side.

dokterbob replied:
Thanks for the super quick feedback by the way. ❤️

0cc4m replied:
> I considered that, but then again if these features do not cause problems on other platforms, and it yields a more generic and simple solution, shouldn't we just leave them on?
> 
> Your call. Curious to see whether you're running into any issues on your side.

It doesn't cause any issues for me, but it might on other systems. What you're using is an instance extension, and it will not be available on all devices. There's also no point enabling it on any other system, since the Vulkan drivers on Linux and Windows are conformant, unlike MoltenVK.

Even on Mac the correct way is checking whether that extension is available before trying to load it. Do you want to add these checks yourself or would you prefer me to do that?

dokterbob replied:
I don't have the foggiest how to add these checks and don't really have bandwitdh to figure it out, honestly.

But will happily wrap my patch with `#if defined (__APPLE__)` (not sure if that's the appropriate but that's what I found).

Very very very happy with the results by the way. Very stable, I now generate without thermal throttling, which promises more gains in the near future! :)

dokterbob replied:
I see the exact patch necessary seems to be in #5322. Happy to blend the patches and update in the next couple of days. Or you can just merge that one if you'd like to move ahead.

0cc4m replied:
Please do, yeah. If you're not sure how to check for the instance availability, I'll add that part after you did the Mac checks.

philiptaron commented:
Could you edit the flake (see https://github.com/ggerganov/llama.cpp/commit/60ecf099eddfe70fec797ef6790572e452054add) and see if you can build with Vulkan on x86_64 MacOS?

dokterbob committed:
Add preprocessor checks for Apple devices.

Based on work by @rbourgeat in https://github.com/ggerganov/llama.cpp/pull/5322/files

dokterbob commented:
> Could you edit the flake (see [60ecf09](https://github.com/ggerganov/llama.cpp/commit/60ecf099eddfe70fec797ef6790572e452054add)) and see if you can build with Vulkan on x86_64 MacOS?

Does that imply removing `useVulkan` on this line?
https://github.com/ggerganov/llama.cpp/commit/60ecf099eddfe70fec797ef6790572e452054add#diff-ed99c076a4accf9ff1d7ad65a79dc3802df6bc8c91c9879e12c766a1b414272bR262

slp commented:
This should probably have it's own issue, but I've tested this PR on an Apple Silicon M1 and noticed I was getting garbage output to my prompts when `batch_size != 1`. Rebuilding with `LLAMA_VULKAN_RUN_TESTS=1` I found that only M-sized matmuls work properly, both S-sized and L-sized fail. Here's an example:

```
TEST F16_F32_S m=100 n=46 k=558 batch=2 split_k=1 matmul 4.609ms avg_err=6.90319
m = 32 n = 0 b = 0
Actual result:

               0       1       2       3       4       5       6       7       8       9
     27:    4.73   12.32    1.43    6.25   -2.21    7.54    6.00    1.09   13.29    6.43
     28:   -2.47    1.54   11.37    7.52  -15.77   11.56   -1.01   -2.09   -0.30  -11.93
     29:   -7.33   -4.47   -4.05   -6.62   -3.47   -7.88    0.44   -5.68  -15.74   -0.46
     30:    6.85    3.08    4.87    5.17    4.22    6.31  -12.52   -3.55   -4.75    2.67
     31:  -16.09    3.49   -8.56   -8.89   -0.33    6.08  -13.39    5.76  -13.02   -8.90
     32:    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
     33:    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
     34:    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
     35:    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
     36:    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00
Expected result:

               0       1       2       3       4       5       6       7       8       9
     27:    4.73   12.32    1.43    6.25   -2.21    7.54    6.00    1.09   13.29    6.43
     28:   -2.47    1.54   11.37    7.52  -15.77   11.56   -1.01   -2.09   -0.30  -11.93
     29:   -7.33   -4.47   -4.05   -6.62   -3.47   -7.88    0.44   -5.68  -15.74   -0.46
     30:    6.85    3.08    4.87    5.17    4.22    6.31  -12.52   -3.55   -4.75    2.67
     31:  -16.09    3.49   -8.56   -8.89   -0.33    6.08  -13.39    5.76  -13.02   -8.90
     32:   12.42    6.19   -2.81    9.91   -9.84    1.53   -5.20   -0.57   -5.79   -4.55
     33:   -7.44   -0.05   12.08   -1.73    5.63   -6.50   -9.00  -17.33   -5.36   -9.06
     34:   -3.51    0.46   -5.92   -3.95   -4.04   15.13   -1.47   -2.21    0.37   10.94
     35:    8.12    1.64  -13.92    3.49    7.47    8.52   -4.61   16.53    1.28   -4.64
     36:   -3.86   -9.55    7.20   -3.38    4.28    2.46   11.96   -4.71    0.67   16.10
```

@0cc4m is this something expected with integrated GPUs? Can I help debug this further somehow?

dokterbob commented:
> > > Could you edit the flake (see [60ecf09](https://github.com/ggerganov/llama.cpp/commit/60ecf099eddfe70fec797ef6790572e452054add)) and see if you can build with Vulkan on x86_64 MacOS?
> > 
> > 
> > Does that imply removing `useVulkan` on this line? [60ecf09#diff-ed99c076a4accf9ff1d7ad65a79dc3802df6bc8c91c9879e12c766a1b414272bR262](https://github.com/ggerganov/llama.cpp/commit/60ecf099eddfe70fec797ef6790572e452054add#diff-ed99c076a4accf9ff1d7ad65a79dc3802df6bc8c91c9879e12c766a1b414272bR262)
> 
> Yeah, just remove all vulkan references from broken or badPlatform.
> 
> To install nix, follow https://github.com/DeterminateSystems/nix-installer. Then `nix build github:dokterbob/llama.cpp/macos_vulkan`. You can uninstall Nix afterwards if you don't want it around.

I can make requested patches today but testing them locally would take me longer, ETA ~week.

Is there a way we can get this in sooner? For Intel Mac users this really is a huge improvement.

0cc4m commented:
> > > > Could you edit the flake (see [60ecf09](https://github.com/ggerganov/llama.cpp/commit/60ecf099eddfe70fec797ef6790572e452054add)) and see if you can build with Vulkan on x86_64 MacOS?
> > > 
> > > 
> > > Does that imply removing `useVulkan` on this line? [60ecf09#diff-ed99c076a4accf9ff1d7ad65a79dc3802df6bc8c91c9879e12c766a1b414272bR262](https://github.com/ggerganov/llama.cpp/commit/60ecf099eddfe70fec797ef6790572e452054add#diff-ed99c076a4accf9ff1d7ad65a79dc3802df6bc8c91c9879e12c766a1b414272bR262)
> > 
> > 
> > Yeah, just remove all vulkan references from broken or badPlatform.
> > To install nix, follow https://github.com/DeterminateSystems/nix-installer. Then `nix build github:dokterbob/llama.cpp/macos_vulkan`. You can uninstall Nix afterwards if you don't want it around.
> 
> I can make requested patches today but testing them locally would take me longer, ETA ~week.
> 
> Is there a way we can get this in sooner? For Intel Mac users this really is a huge improvement.

I can try doing the nix tests, I think. I'll also add code to check if the extension is available. It would be good if you could test that afterwards, since I don't have a Mac system. Then we should be able to merge. Would that work?

dokterbob commented:
> I can try doing the nix tests, I think. I'll also add code to check if the extension is available. It would be good if you could test that afterwards, since I don't have a Mac system. Then we should be able to merge. Would that work?

Sure, happy to test!

0cc4m committed:
Add check for VK_KHR_portability_enumeration for MoltenVK support

0cc4m committed:
Enable Vulkan MacOS CI

0cc4m commented:
> > I can try doing the nix tests, I think. I'll also add code to check if the extension is available. It would be good if you could test that afterwards, since I don't have a Mac system. Then we should be able to merge. Would that work?
> 
> Sure, happy to test!

@dokterbob I added the checks. Can you check if it works on Apple?

I tried build it with nix locally and it didn't throw an error, but I have no idea how nix works or how to use it. @philiptaron Let me know if it looks correct. The CI should fail if it doesn't work, right?

philiptaron commented:
I think you also need this change:

```diff
diff --git a/flake.nix b/flake.nix
index ad2f9b29..dc4e503c 100644
--- a/flake.nix
+++ b/flake.nix
@@ -150,6 +150,7 @@
             packages =
               {
                 default = config.legacyPackages.llamaPackages.llama-cpp;
+                vulkan = config.packages.default.override { useVulkan = true; };
               }
               // lib.optionalAttrs pkgs.stdenv.isLinux {
                 opencl = config.packages.default.override { useOpenCL = true; };
@@ -157,7 +158,6 @@

                 mpi-cpu = config.packages.default.override { useMpi = true; };
                 mpi-cuda = config.packages.default.override { useMpi = true; };
-                vulkan = config.packages.default.override { useVulkan = true; };
               }
               // lib.optionalAttrs (system == "x86_64-linux") {
                 rocm = config.legacyPackages.llamaPackagesRocm.llama-cpp;
```

0cc4m commented:
> I think you also need this change:
> 
> ```diff
> diff --git a/flake.nix b/flake.nix
> index ad2f9b29..dc4e503c 100644
> --- a/flake.nix
> +++ b/flake.nix
> @@ -150,6 +150,7 @@
>              packages =
>                {
>                  default = config.legacyPackages.llamaPackages.llama-cpp;
> +                vulkan = config.packages.default.override { useVulkan = true; };
>                }
>                // lib.optionalAttrs pkgs.stdenv.isLinux {
>                  opencl = config.packages.default.override { useOpenCL = true; };
> @@ -157,7 +158,6 @@
> 
>                  mpi-cpu = config.packages.default.override { useMpi = true; };
>                  mpi-cuda = config.packages.default.override { useMpi = true; };
> -                vulkan = config.packages.default.override { useVulkan = true; };
>                }
>                // lib.optionalAttrs (system == "x86_64-linux") {
>                  rocm = config.legacyPackages.llamaPackagesRocm.llama-cpp;
> ```

You can also push commits to this PR. I think it's better if you do that, I have no clue what I'm doing in the nix files.

philiptaron commented:
> You can also push commits to this PR. I think it's better if you do that, I have no clue what I'm doing in the nix files.

It might be significantly delayed then; it's a week of real-life responsibilities for me. 🙂

dokterbob committed:
nix: now that we can do so, allow MacOS to build Vulkan binaries

Author:    Philip Taron <philip.taron@gmail.com>
Date:      Tue Feb 13 20:28:02 2024 +0000

philiptaron commented on ggml-vulkan.cpp:
```diff
@@ -1088,17 +1088,45 @@ void ggml_vk_instance_init() {
-    const std::vector<const char*> layers = {
+
+    const std::vector<vk::ExtensionProperties> instance_extensions = vk::enumerateInstanceExtensionProperties();
+#ifdef __APPLE__
```
Nit: I would prefer that these get extracted to functions that either do something (in the Apple case) or do nothing (in the other cases) and land at the bottom of this file; see `ggml.c` for an example with `__APPLE__` around line 16583.

0cc4m replied:
You're right. I reworked this and the validation layer enable. Can you take a look (and maybe do another sanity check run on Apple)?

philiptaron reviewed (approved):
I built this on Sonoma, both Intel Mac and M2 Mac. Compiles and runs. If @ggerganov doesn't object, I aim to merge this in 12 hours.

0cc4m committed:
Refactor validation and enumeration platform checks into functions to clean up ggml_vk_instance_init()

philiptaron commented on ggml-vulkan.cpp:
```diff
@@ -1088,28 +1091,42 @@ void ggml_vk_instance_init() {
+    const bool validation_ext = ggml_vk_instance_validation_ext_available(instance_extensions);
+    const bool portability_enumeration_ext = ggml_vk_instance_portability_enumeration_ext_available(instance_extensions);
+
+    std::vector<const char*> layers;
+
+    if (validation_ext) {
+        layers.push_back("VK_LAYER_KHRONOS_validation");
+    }
+    std::vector<const char*> extensions;
+    if (validation_ext) {
+        extensions.push_back("VK_EXT_validation_features");
+    }
+    if (portability_enumeration_ext) {
+        extensions.push_back("VK_KHR_portability_enumeration");
+    }
+    vk::InstanceCreateInfo instance_create_info(vk::InstanceCreateFlags{}, &app_info, layers, extensions);
+    if (portability_enumeration_ext) {
+        instance_create_info.flags |= vk::InstanceCreateFlagBits::eEnumeratePortabilityKHR;
+    }
+
+    std::vector<vk::ValidationFeatureEnableEXT> features_enable;
+    vk::ValidationFeaturesEXT validation_features;
+
+    if (validation_ext) {
+        features_enable = { vk::ValidationFeatureEnableEXT::eBestPractices };
+        validation_features = {
+            features_enable,
+            {},
+        };
+        validation_features.setPNext(nullptr);
+        instance_create_info.setPNext(&validation_features);
+
+        std::cerr << "ggml_vulkan: Validation layers enabled" << std::endl;
+    }
```
I was hoping that these constants would have one home:

`"VK_LAYER_KHRONOS_validation"`
`"VK_EXT_validation_features"`
`"VK_KHR_portability_enumeration"`

Could you accomplish that by making a function that returned or mutated the two variables `features_enable` and `validation_features`?

That way, this generic code could stay unaware of the sort of features and validation that got added, and the system-specific code below would be able to add (as needed) the right features and validation.

0cc4m replied:
I could, but let's not overextend the scope and duration of this PR. If you wanna do a refactor PR yourself or in collaboration with me, or gather suggestions in an issue, that's fine.

philiptaron replied:
I went to try to do that refactoring with a spare hour this morning, and ran into a rebase error (on 8b38eb59) that I could not see how to resolve in a straightforward manner. Could you or @dokterbob rebase this branch on master appropriately?

0cc4m replied:
I didn't have any issues merging upstream into the branch, not sure what your problem was.

To deal with the linker error that Vulkan currently has with cmake, you can use this workaround patch:
```patch
diff --git a/examples/gguf/CMakeLists.txt b/examples/gguf/CMakeLists.txt
index 6481f087..85950145 100644
--- a/examples/gguf/CMakeLists.txt
+++ b/examples/gguf/CMakeLists.txt
@@ -1,5 +1,5 @@
 set(TARGET gguf)
 add_executable(${TARGET} gguf.cpp)
 install(TARGETS ${TARGET} RUNTIME)
-target_link_libraries(${TARGET} PRIVATE ggml ${CMAKE_THREAD_LIBS_INIT})
+target_link_libraries(${TARGET} PRIVATE common ggml ${CMAKE_THREAD_LIBS_INIT})
 target_compile_features(${TARGET} PRIVATE cxx_std_11)
```

philiptaron reviewed (changes requested):
I took a look at the code; I'd like to see the refactor tractor continue to plow, @0cc4m. 😀

philiptaron commented:
I intend to merge after CI completes.

ggerganov commented:
Keep in mind to squash the commits when merging in the future 👍

philiptaron commented:
> Keep in mind to squash the commits when merging in the future 👍

If that's the style preferred in this repo, I'll make sure to do that next time.

thewh1teagle commented:
Any chance we should enable it on linux/windows as well? I'm getting similar error with whisper.cpp + Vulkan.
